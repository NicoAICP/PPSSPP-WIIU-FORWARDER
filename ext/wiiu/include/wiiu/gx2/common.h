#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
