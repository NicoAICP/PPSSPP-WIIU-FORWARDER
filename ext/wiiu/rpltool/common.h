#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
