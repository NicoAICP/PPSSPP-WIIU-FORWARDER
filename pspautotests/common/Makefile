TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
