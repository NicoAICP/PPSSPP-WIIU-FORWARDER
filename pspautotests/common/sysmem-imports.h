extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();