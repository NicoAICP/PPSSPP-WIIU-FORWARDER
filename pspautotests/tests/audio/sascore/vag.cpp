#include <common.h>
#include <psputility.h>
#include <pspiofilemgr.h>
#include "sascore.h"

__attribute__((aligned(64))) SasCore sasCore;
__attribute__((aligned(64))) short pcm[256 * 2 * 16] = {0};
__attribute__((aligned(64))) short samples[4096 * 2 * 16] = {0};
u8 *vag;

void testSetVoice(SasCore *sasCore, int v, void *addr, int size, int loop, const char *title) {
	int result = __sceSasSetVoice(sasCore, v, addr, size, loop);
	if (result < 0) {
		checkpoint("%s: Failed (%08x)", title, result);
	} else {
		checkpoint("%s: OK (%08x)", title, result);
	}
}

void printSample(int i) {
	schedf("  [%03x] %04x %04x\n", i, (u16)samples[i * 2 + 0], (u16)samples[i * 2 + 1]);
}

void testVAGOutput(int loop) {
	memset(samples, 0xdd, sizeof(samples));

	__sceSasSetKeyOff(&sasCore, 0);
	__sceSasSetADSRmode(&sasCore, 0, 15, PSP_SAS_ADSR_CURVE_MODE_LINEAR_INCREASE, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT);
	__sceSasSetADSR(&sasCore, 0, 15, 0x40000000, 0x40000000, 0x40000000, 0x40000000);

	char temp[64];
	snprintf(temp, sizeof(temp), "Output with loop %04x:", loop);
	checkpointNext(temp);
	__sceSasSetGrain(&sasCore, 512);
	__sceSasSetOutputmode(&sasCore, 0);
	__sceSasSetVoice(&sasCore, 0, vag, 1024, loop);
	__sceSasSetKeyOn(&sasCore, 0);
	memset(samples, 0xdd, sizeof(samples));
	__sceSasCore(&sasCore, samples);
	printSample(0x0020);
	printSample(0x011E);
	printSample(0x0120);
	checkpoint("  Ended: %08x", __sceSasGetEndFlag(&sasCore));
}

void testVAGData(int predict_nr, int flagsAfter0) {
	u8 data[0x100];

	// Each block is 16 bytes, predict_nr, flags, data.
	for (int i = 0; i < 0x100; i += 0x10) {
		data[i + 0] = (predict_nr << 4) | (i >> 4);
		data[i + 1] = i == 0 ? 0 : flagsAfter0;
		for (int j = 0x02; j < 0x10; ++j) {
			data[i + j] = (j << 4) | 3;
		}
	}

	__sceSasSetKeyOff(&sasCore, 0);
	__sceSasSetADSRmode(&sasCore, 0, 15, PSP_SAS_ADSR_CURVE_MODE_LINEAR_INCREASE, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT);
	__sceSasSetADSR(&sasCore, 0, 15, 0x40000000, 0x40000000, 0x40000000, 0x40000000);

	char temp[64];
	snprintf(temp, sizeof(temp), "VAG data (predict_nr=%x, flagsAfter0=%02x):", predict_nr, flagsAfter0);
	checkpointNext(temp);
	__sceSasSetGrain(&sasCore, 512);
	__sceSasSetOutputmode(&sasCore, 0);
	__sceSasSetVoice(&sasCore, 0, data, 0x100, 1);
	__sceSasSetKeyOn(&sasCore, 0);
	memset(samples, 0xdd, sizeof(samples));
	__sceSasCore(&sasCore, samples);
	for (int i = 32; i < 40; ++i) {
		printSample(i);
	}
	checkpoint("  Ended: %08x", __sceSasGetEndFlag(&sasCore));
}

extern "C" int main(int argc, char *argv[]) {
	sceUtilityLoadModule(PSP_MODULE_AV_AVCODEC);
	sceUtilityLoadModule(PSP_MODULE_AV_SASCORE);

	int fd = sceIoOpen("music.vag", PSP_O_RDONLY, 0);
	int sz = sceIoLseek32(fd, 0, PSP_SEEK_END);
	sceIoLseek32(fd, 0, SEEK_SET);
	vag = new u8[sz];
	sceIoRead(fd, vag, sz);
	sceIoClose(fd);

	__sceSasInit(&sasCore, 128, 32, 1, 44100);

	for (int i = 0; i < ARRAY_SIZE(pcm); ++i) {
		pcm[i] = (short)(65535 - i);
	}

	checkpointNext("Voices:");
	testSetVoice(&sasCore, 0, vag, 0x100, 0, "  Normal");
	testSetVoice(&sasCore, -1, vag, 0x100, 0, "  Voice -1");
	testSetVoice(&sasCore, 32, vag, 0x100, 0, "  Voice 32");

	checkpointNext("Addresses:");
	testSetVoice(&sasCore, 0, 0, 0x100, 0, "Zero");
	testSetVoice(&sasCore, 0, vag, 0x100, 0, "Valid");
	
	checkpointNext("Sizes:");
	static const int sizes[] = {-500000, -64, -48, -47, -1, 0, 1, 2, 3, 4, 5, 8, 12, 15, 16, 32, 48, 64, 128, 255, 256, 257, 1024, 1025, 0x10000, 0x10001, 0x20000, 0x10000000, 0x80000000};
	for (int i = 0; i < ARRAY_SIZE(sizes); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Size %x", sizes[i]);
		testSetVoice(&sasCore, 0, vag, sizes[i], 0, temp);
	}

	checkpointNext("Loop start pos:");
	static const int loops[] = {-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 255, 256, 257, 0x10000, 0x10001, 0x10000000, 0x40000001, 0x80000001};
	for (int i = 0; i < ARRAY_SIZE(loops); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Loop pos %d / 256", loops[i]);
		testSetVoice(&sasCore, 0, vag, 0x100, loops[i], temp);
	}
	testSetVoice(&sasCore, 0, vag, 1, 1, "  Loop pos 1 / 1");
	testSetVoice(&sasCore, 0, vag, 1, 2, "  Loop pos 2 / 1");
	testSetVoice(&sasCore, 0, vag, 0x10000, 0xFFFF, "  Loop pos 0xFFFF / 0x10000");
	testSetVoice(&sasCore, 0, vag, 0x10000, 0x10000, "  Loop pos 0x10000 / 0x10000");

	__sceSasSetKeyOff(&sasCore, 0);
	__sceSasSetVoicePCM(&sasCore, 0, pcm, 256, 0);
	__sceSasSetADSRmode(&sasCore, 0, 15, PSP_SAS_ADSR_CURVE_MODE_LINEAR_INCREASE, PSP_SAS_ADSR_CURVE_MODE_LINEAR_DECREASE, PSP_SAS_ADSR_CURVE_MODE_LINEAR_DECREASE, PSP_SAS_ADSR_CURVE_MODE_LINEAR_DECREASE);
	__sceSasSetSL(&sasCore, 0, 0x12345678);
	__sceSasSetADSR(&sasCore, 0, 15, 0x1000, 0x1000, 0x1000, 0x1000);

	checkpointNext("After voice setup:");
	testSetVoice(&sasCore, 0, vag, 0x10000, 0, "  Normal");
	__sceSasSetKeyOff(&sasCore, 0);
	__sceSasSetPause(&sasCore, -1, 1);
	testSetVoice(&sasCore, 0, vag, 0x10000, 0, "  While paused");
	checkpoint("  Still paused: %08x", __sceSasGetPauseFlag(&sasCore));
	__sceSasSetPause(&sasCore, -1, 0);
	
	checkpointNext("With core:");
	checkpoint("  Key on: %08x", __sceSasSetKeyOn(&sasCore, 0));
	// Do it a few times to bump the height up.
	for (int i = 0; i < 4; ++i) {
		checkpoint("  Core: %08x", __sceSasCore(&sasCore, samples));
	}
	testSetVoice(&sasCore, 0, vag, 0x10000, 0, "  After processing");
	checkpoint("  Still playing: %08x", __sceSasGetEndFlag(&sasCore));

	// TODO: Do we want to emulate the resampling exactly?
	testVAGOutput(0);
	testVAGOutput(1);

	delete [] vag;

	for (int i = 0; i < 0x10; ++i) {
		testVAGData(i, 0);
	}
	testVAGData(0x0, 0x03);
	testVAGData(0x0, 0x07);
	testVAGData(0x0, 0x41);
	testVAGData(0x0, 0x01);
	testVAGData(0x0, 0x87);

	return 0;
}
