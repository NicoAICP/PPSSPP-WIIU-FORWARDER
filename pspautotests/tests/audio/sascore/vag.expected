[x] Voices:
[x]   Normal: OK (00000000)
[x]   Voice -1: Failed (80420010)
[x]   Voice 32: Failed (80420010)

[x] Addresses:
[x] Zero: OK (00000000)
[x] Valid: OK (00000000)

[x] Sizes:
[x]   Size fff85ee0: OK (00000000)
[x]   Size ffffffc0: OK (00000000)
[x]   Size ffffffd0: OK (00000000)
[x]   Size ffffffd1: Failed (80420014)
[x]   Size ffffffff: Failed (80420014)
[x]   Size 0: Failed (80420014)
[x]   Size 1: Failed (80420014)
[x]   Size 2: Failed (80420014)
[x]   Size 3: Failed (80420014)
[x]   Size 4: Failed (80420014)
[x]   Size 5: Failed (80420014)
[x]   Size 8: Failed (80420014)
[x]   Size c: Failed (80420014)
[x]   Size f: Failed (80420014)
[x]   Size 10: OK (00000000)
[x]   Size 20: OK (00000000)
[x]   Size 30: OK (00000000)
[x]   Size 40: OK (00000000)
[x]   Size 80: OK (00000000)
[x]   Size ff: Failed (80420014)
[x]   Size 100: OK (00000000)
[x]   Size 101: Failed (80420014)
[x]   Size 400: OK (00000000)
[x]   Size 401: Failed (80420014)
[x]   Size 10000: OK (00000000)
[x]   Size 10001: Failed (80420014)
[x]   Size 20000: OK (00000000)
[x]   Size 10000000: OK (00000000)
[x]   Size 80000000: OK (00000000)

[x] Loop start pos:
[x]   Loop pos -1 / 256: Failed (80420015)
[x]   Loop pos 0 / 256: OK (00000000)
[x]   Loop pos 1 / 256: OK (00000000)
[x]   Loop pos 2 / 256: Failed (80420015)
[x]   Loop pos 3 / 256: Failed (80420015)
[x]   Loop pos 4 / 256: Failed (80420015)
[x]   Loop pos 5 / 256: Failed (80420015)
[x]   Loop pos 6 / 256: Failed (80420015)
[x]   Loop pos 7 / 256: Failed (80420015)
[x]   Loop pos 8 / 256: Failed (80420015)
[x]   Loop pos 255 / 256: Failed (80420015)
[x]   Loop pos 256 / 256: Failed (80420015)
[x]   Loop pos 257 / 256: Failed (80420015)
[x]   Loop pos 65536 / 256: Failed (80420015)
[x]   Loop pos 65537 / 256: Failed (80420015)
[x]   Loop pos 268435456 / 256: Failed (80420015)
[x]   Loop pos 1073741825 / 256: Failed (80420015)
[x]   Loop pos -2147483647 / 256: Failed (80420015)
[x]   Loop pos 1 / 1: Failed (80420014)
[x]   Loop pos 2 / 1: Failed (80420014)
[x]   Loop pos 0xFFFF / 0x10000: Failed (80420015)
[x]   Loop pos 0x10000 / 0x10000: Failed (80420015)

[x] After voice setup:
[x]   Normal: OK (00000000)
[x]   While paused: OK (00000000)
[x]   Still paused: ffffffff

[x] With core:
[x]   Key on: 00000000
[r]   Core: 00000000
[r]   Core: 00000000
[r]   Core: 00000000
[r]   Core: 00000000
[x]   After processing: OK (00000000)
[x]   Still playing: fffffffe

[x] Output with loop 0000:
  [020] 0000 0000
  [11e] 2d4b 2d4b
  [120] 2cb4 2cb4
[r]   Ended: fffffffe

[x] Output with loop 0001:
  [020] 0000 0000
  [11e] 2d4b 2d4b
  [120] 2cb4 2cb4
[r]   Ended: fffffffe

[x] VAG data (predict_nr=0, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=1, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 4d00 4d00
  [023] 7830 7830
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=2, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 7640 7640
  [023] 7fff 7fff
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=3, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 6980 6980
  [023] 7fff 7fff
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=4, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 7b80 7b80
  [023] 7fff 7fff
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=5, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=6, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=7, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 4700 4700
  [023] 69b0 69b0
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=8, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 4940 4940
  [023] 6d73 6d73
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=9, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 4d00 4d00
  [023] 1a70 1a70
  [024] b265 b265
  [025] b39b b39b
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=a, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=b, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] ebc0 ebc0
  [024] 0280 0280
  [025] 4ccb 4ccb
  [026] 3c72 3c72
  [027] c2cf c2cf
[r]   Ended: ffffffff

[x] VAG data (predict_nr=c, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=d, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2180 2180
  [023] 8f0c 8f0c
  [024] bb68 bb68
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 8000 8000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=e, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 7dc0 7dc0
  [023] 7fff 7fff
  [024] 7fff 7fff
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 7fff 7fff
[r]   Ended: ffffffff

[x] VAG data (predict_nr=f, flagsAfter0=00):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] bec0 bec0
  [024] e480 e480
  [025] 7fff 7fff
  [026] 7fff 7fff
  [027] 8000 8000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=0, flagsAfter0=03):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: fffffffe

[x] VAG data (predict_nr=0, flagsAfter0=07):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=0, flagsAfter0=41):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=0, flagsAfter0=01):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff

[x] VAG data (predict_nr=0, flagsAfter0=87):
  [020] 0000 0000
  [021] 3000 3000
  [022] 2000 2000
  [023] 3000 3000
  [024] 3000 3000
  [025] 3000 3000
  [026] 4000 4000
  [027] 3000 3000
[r]   Ended: ffffffff
